src/CodeGen_Hexagon.cpp

```diff
@@ -0,0 +1,1382 @@
+
+using std::vector;
+using std::string;
+
+using namespace llvm;
+
+CodeGen_Hexagon::CodeGen_Hexagon(Target t) : CodeGen_Posix(t) {}
+
+std::unique_ptr<llvm::Module> CodeGen_Hexagon::compile(const Module &module) {
+    auto llvm_module = CodeGen_Posix::compile(module);
+    static bool options_processed = false;
+
+    // TODO: This should be set on the module itself, or some other
+    // safer way to pass this through to the target specific lowering
+    // passes. We set the option here (after the base class'
+    // implementaiton of compile) because it is the last
+    // Hexagon-specific code to run prior to invoking the target
+    // specific lowering in LLVM, minimizing the chances of the wrong
+    // flag being set for the wrong module.
+    if (!options_processed) {
+        cl::ParseEnvironmentOptions("halide-hvx-be", "HALIDE_LLVM_ARGS",
+                                    "Halide HVX internal compiler\n");
+
+        std::vector<const char *> options = {
+            // Don't put small objects into .data sections, it causes
+            // issues with position independent code.
+            "-hexagon-small-data-threshold=0"
+        };
+        cl::ParseCommandLineOptions(options.size(), options.data());
+    }
+    options_processed = true;
+
+    if (module.target().features_all_of({Halide::Target::HVX_128, Halide::Target::HVX_64})) {
+        internal_error << "Both HVX_64 and HVX_128 set at same time\n";
+    }
+
+    return llvm_module;
+}
+
+namespace {
+
+// A piece of IR uses HVX if it contains any vector type producing IR
+// nodes.
+class UsesHvx : public IRVisitor {
+private:
+    using IRVisitor::visit;
+    void visit(const Variable *op) {
+        uses_hvx = uses_hvx || op->type.is_vector();
+    }
+    void visit(const Ramp *op) {
+        uses_hvx = uses_hvx || op->type.is_vector();
+    }
+    void visit(const Broadcast *op) {
+        uses_hvx = uses_hvx || op->lanes > 1;
+    }
+    void visit(const Call *op) {
+        uses_hvx = uses_hvx || op->type.is_vector();
+    }
+
+public:
+    bool uses_hvx = false;
+};
+
+bool uses_hvx(Stmt s) {
+    UsesHvx uses;
+    s.accept(&uses);
+    return uses.uses_hvx;
+}
+
+}  // namespace
+
+void CodeGen_Hexagon::compile_func(const LoweredFunc &f,
+                                   const string &simple_name, const string &extern_name) {
+    CodeGen_Posix::begin_func(f.linkage, simple_name, extern_name, f.args);
+
+    Stmt body = f.body;
+
+    debug(1) << "Optimizing shuffles...\n";
+    body = optimize_hexagon_shuffles(body);
+    debug(2) << "Lowering after optimizing shuffles:\n" << body << "\n\n";
+
+    debug(1) << "Aligning loads for HVX....\n";
+    body = align_loads(body, target.natural_vector_size(Int(8)));
+    body = common_subexpression_elimination(body);
+    body = simplify(body);
+    debug(2) << "Lowering after aligning loads:\n" << body << "\n\n";
+
+    debug(1) << "Forwarding stores across loop iterations...\n";
+    body = loop_carry(body, 16);
```

Comment:
"Forwarding stores" is no longer try. How about:

    debug(1) << "Carrying values across loop iterations..\n";
    // Use at most 16 registers for carrying values